presto-main/src/main/java/com/facebook/presto/operator/BlockEncodingBuffers.java

```diff
@@ -0,0 +1,417 @@
+
+        if (remainingNullsCount > 0) {
+            byte value = 0;
+            int mask = 0b1000_0000;
+            for (int i = 0; i < remainingNullsCount; i++) {
+                value |= remainingNullsFromLastBatch[i] ? mask : 0;
+                mask >>>= 1;
+            }
+
+            // Process the next a few nulls to make up one byte
+            int positionCountForFirstByte = BITS_IN_BYTE - remainingNullsCount;
+            int endPositionOffset = positionsOffset + positionCountForFirstByte;
+            while (offset < endPositionOffset) {
+                int position = positions[offset];
+                value |= block.isNull(position) ? mask : 0;
+                mask >>>= 1;
+                offset++;
+            }
+            nullsBufferContainsNull |= (value != 0);
+            ByteArrayUtils.writeByte(nullsBuffer, nullsBufferIndex++, value);
+        }
+
+        // Process the next BITS_IN_BYTE * n positions. We now have processed (offset - positionsOffset) positions
+        int remainingPositions = batchSize - (offset - positionsOffset);
+        int positionsToEncode = remainingPositions & ~0b111;
+        int endPositionOffset = offset + positionsToEncode;
+        while (offset < endPositionOffset) {
+            byte value = 0;
+            value |= block.isNull(positions[offset]) ? 0b1000_0000 : 0;
+            value |= block.isNull(positions[offset + 1]) ? 0b0100_0000 : 0;
+            value |= block.isNull(positions[offset + 2]) ? 0b0010_0000 : 0;
+            value |= block.isNull(positions[offset + 3]) ? 0b0001_0000 : 0;
+            value |= block.isNull(positions[offset + 4]) ? 0b0000_1000 : 0;
+            value |= block.isNull(positions[offset + 5]) ? 0b0000_0100 : 0;
+            value |= block.isNull(positions[offset + 6]) ? 0b0000_0010 : 0;
+            value |= block.isNull(positions[offset + 7]) ? 0b0000_0001 : 0;
+
+            nullsBufferContainsNull |= (value != 0);
+            ByteArrayUtils.writeByte(nullsBuffer, nullsBufferIndex, value);
+            nullsBufferIndex++;
+            offset += BITS_IN_BYTE;
+        }
+
+        // Process the remaining positions
+        remainingPositions &= 0b111;
+        verify(remainingPositions < BITS_IN_BYTE);
+        remainingNullsCount = 0;
+        while (remainingNullsCount < remainingPositions) {
+            int position = positions[offset++];
+            remainingNullsFromLastBatch[remainingNullsCount++] = block.isNull(position);
+        }
+    }
+
+    private void encodeFalseValuesAsBits(int count)
+    {
+        if (remainingNullsCount + count < BITS_IN_BYTE) {
+            // just put all of this batch to remainingNullsFromLastBatch
+            for (int i = 0; i < count; i++) {
+                remainingNullsFromLastBatch[remainingNullsCount++] = false;
+            }
+            return;
+        }
+
+        // Have to do this before calling encodeRemainingNullsAsBits() because it resets remainingNullsCount
```

Comment:
this comment is incorrect; encodeRemainingNullsAsBits doesn't reset remainingNullsCount; I'd remove this comment altogether